metaseq/scripts/reshard_mp.py

```diff
@@ -0,0 +1,129 @@
+            Shard indices can be included in the path names if the pattern includes `{i}`.
+            (e.g. "checkpoints/opt-2.7b/reshard_no_os_mp8/reshard-model_part-{i}.pt").
+        :param num_output_parts: The number of output model parallel parts.
+        :param eps: A tolerance threshold for the maxinum discrepancy between MP parts.
+    """
+    files = glob(input)
+    N, M = len(files), num_output_parts
+    if N == 0:
+        raise ValueError("The glob pattern doesn't match any model parallel parts.")
+    files = sorted(files, key=lambda x: list(map(int, re.findall(r"\d+", x))))
+    logger.info(f"Found {len(files)} model parallel parts ({files[0]} to {files[-1]})")
+
+    rank0_state_dict = torch.load(files[0], torch.device("cpu"))
+    dim0_shard_regex = re.compile("embed_tokens|ffn_layernorm|fc1|(k|q|v)_proj")
+    dim1_shard_regex = re.compile("(fc2|out_proj).weight")
+    shared_regex = re.compile(
+        "embed_positions|layer_norm|(fc2|out_proj).bias|output_projection|version"
+    )
+
+    unsharded_dict = {}
+    logger.info("Allocating memory for unsharded checkpoint")
+    for key, value in rank0_state_dict["model"].items():
+        d0, d1 = value.size()[0], value.size()[1:]
+        if "qkv" in key:
+            unsharded_dict[key] = value.new_zeros(3, N * d0 // 3, *d1)
+        elif dim0_shard_regex.search(key):
+            unsharded_dict[key] = value.new_zeros(N * d0, *d1)
+        elif dim1_shard_regex.search(key):
+            assert len(d1) > 0
+            unsharded_dict[key] = value.new_zeros(d0, N * d1[0])
+
+    # Iteratively load checkpoints to avoid OOM issues
+    for i, file in enumerate(files):
+        logger.info(f"Merging {file} into unsharded checkpoint")
+        state_dict = torch.load(file, torch.device("cpu"))
+        for key, value in state_dict["model"].items():
+            d0, d1 = value.size()[0], value.size()[1:]
+            if "qkv" in key:
+                # Split and copy QKV weights
+                unsharded_dict[key][:, i * d0 // 3 : (i + 1) * d0 // 3].copy_(
+                    value.view(3, d0 // 3, *d1)
+                )
+            elif dim0_shard_regex.search(key):
+                # Concatenate along dim 0 (e.g. embed_tokens, fc1.weight, fc1.bias)
+                unsharded_dict[key][i * d0 : (i + 1) * d0].copy_(value)
+            elif dim1_shard_regex.search(key):
+                # Concatenate along dim 1 (e.g. fc2.weight, out_proj.weight)
+                unsharded_dict[key][:, i * d1[0] : (i + 1) * d1[0]].copy_(value)
+            elif shared_regex.search(key):
+                # Copy from rank 0 (e.g. embed_positions, final_layer_norm, fc2.bias, out_proj.bias)
+                unsharded_dict[key] = value
+                diff = _max_diff(rank0_state_dict["model"][key], value)
+                if diff > eps:
+                    logger.warning(f"Max value discrepancy for key '{key}': {diff:.4e}")
```

Comment:
Throw an error instead? Warnings can be easily missed

Reply:
This is actually expected to happen quite often with layer norm biases (see the logs P587377365 for example), so I think we can just follow the previous script and keep the warning here.